### ISSUE TITLE
Access denied in call

### ISSUE THREAD

**Issue by erviem99** [digst/OIOIDWS.Net#41]
Hi!

I'm trying to modify Digst.OioIdws.Rest.Example.Client to return an access token for thelookupservices described here, but I'm not having too much luck: https://cms.nemlog-in.dk/media/tb4l3i0i/nl3-supporting-services-v1-7.pdf

I modified the AsEndpoint (Access Token Issuer Endpoint) in the code to point to: https://lookupservice.test-devtest4-nemlog-in.dk/api/accesstoken/issue

I will get a saml-token back, but when I want to exchange that for an access-token I just get an "Access Denied" (Invalid token) back.

What is it that I am missing?

Regards,
Erik

**Comment by erviem99** [comment]
The integration tests also return either "SOAP signature recieved from STS does not validate!" or "Authentication failed".

Something is clearly wrong.

**Comment by martinwiboe** [comment]
@erviem99  did you get this to work? I can't even get STS to return a SAML token.

**Comment by erviem99** [comment]
Unfortunately not. It seems like it mostly broke around June 20, and is hasn't been working since.

I can see there is a service-window later this week and next week and I pin all my hopes on that now.

Communication from DIGST NemLog-in is almost non-existant (I've been trying for months now).

**Comment by martinwiboe** [comment]
This state of affairs is beyond sad -- NemLog-in is really a foundation of Danish public infrastructure.

I have replaced the client certificate with the one from https://www.nemlog-in.dk/vejledningertiltestmiljo/forside/test-som-tjenesteudbyder-eller-broker/ (in the Opslagstjenester section) and I have changed the AudienceUri accordingly to `https://saml.wsp.lookupservice.test-devtest4-nemlog-in.dk`. But I simply get a "Malformed request" error from STS.

I can see the request in the Log Viewer, but it fails with some exception on their end, and it's not clear why.  
NemLog-in Support is not getting back to me.

Were you able to get a token from STS?

**Comment by erviem99** [comment]
I could not agree more.

It is also shocking that we are still using Ws-Trust protocol. Because of that we are stuck on the old .Net Framework.

I did the exact same thing as you without luck (and a million other combinations). 

The only situation where I can get a SAML token back is by running the Digst.OioIdws.Rest.Examples.Client example.

Unfortunatly, I need the lookupservice just like you and not some test-service.

**Comment by martinwiboe** [comment]
Yes - once I get the sample to work, I will try to implement the WS-Trust request in .NET 6. Apparently parts of WCF are ported to the platform, but my first approach will be to construct the SOAP request manually. The code by [ITfoxtec](https://github.com/ITfoxtec/ITfoxtec.Identity.Saml2) will serve as an inspiration in this regard.

I will update this thread if I make any progress.

**Comment by martinwiboe** [comment]
Also pinging @mdbech in case you have any input?

**Comment by erviem99** [comment]
I would love to see working code for Net6 that will also run under Linux. Please keep me posted :-)

**Comment by martinwiboe** [comment]
After running the code with the original certificate, but setting the AS endpoint to the one at test-devtest4, it looks to me as if the client is sending the EncryptedAssertion from STS directly to the AS. Won't this cause problems? I thought the encryption was done with the public key belonging to the client and thus unreadable for the AS.

**Comment by erviem99** [comment]
I don't think there is anything wrong with the code as I have had it working before with calls to lookupservice (in devtest4). It just stopped working around june 20, perhaps when the certificates was updated.

**Comment by erviem99** [comment]
@martinwiboe Any luck? Service Window is over but (of course) no information at all from Digst. It still does not work for me.

**Comment by martinwiboe** [comment]
I haven't had any luck. I received a response from NemLog-in Support, and I've asked them to look at this sample and why it's not working (rather than try to troubleshoot with my own certificate.) If they respond, I will update this thread.

I also believe that you are correct that there are no issues with the code. The STS response will be encrypted with the public key of the WSP (ie. the Authorization Service) and as such should not be decrypted by the consuming client.

Pinging @CBechmann

**Comment by erviem99** [comment]
I mangel på bedre tænkte jeg at vi måske kunne hjælpe hinanden lidt videre her...

Det er lykkes for mig at få et access-token tilbage fra STS ved at signere med et brugercertifikat udstedt herfra: https://erhvervsadministration.test-devtest4-nemlog-in.dk/. Man bestiller et nyt certifikat under punktet Brugercertifikat.

Der hvor jeg nå går fast er når jeg skal oprette en systembruger og tildele den privilegier til Lookup Service. Dette gøres under https://administration.devtest4-nemlog-in.dk/.

Jeg har gjort alt dette og fået en mail tilbage om at "Anmodningen er blevet godkendt", men desværre får jeg stadig en 403 (Forbidden) i kaldet til Lookup Service.

Der hvor kæden hopper af for mig er dog hvordan mit brugercertifikat kobles sammen med systembrugeren i Lookup Servicen. Det kan jeg ikke se nogen steder hvordan foregår. 

Jeg ved selvfølgelig ikke om det er derfor jeg får en 403 tilbage...

**Comment by martinwiboe** [comment]
Det er en god idé. Jeg har modtaget svar fra NemLog-in support, men ikke noget brugbart -- jeg har bedt dem om at give anvisninger i at få sample til at virke.

Sker sammenkoblingen ikke via EntityId?

Det kan være, at jeg får tid til at prøve samme fremgangsmåde som dig i weekenden. Så vender jeg tilbage (og selvfølgelig også, hvis jeg kommer videre med supporten).

**Comment by erviem99** [comment]
Jeg er i tvivl, fordi EntityId indgår ikke i selve kaldet til Lookup Services (eks, api/lookup/ridcpr).

Der er dog en konfiguration på selve systembrugeren, hvor man skal indsætte et VOCES eller et FOCES certifikat og som benyttes til selve signeringen.

Jeg ved ikke helt hvor det certifikat skal komme fra, men hvis man har et IT-system sat op i administration.devtest4-nemlog-in.dk, så kan man downloade et signeringscertifikat (.cer - fil) fra siden for IT-systemet som fungerer.

Min udfordring er dog at jeg ikke har noget IT-system sat op til min nye testorganisation, og jeg ved ikke hvordan jeg gør dette.

Jeg satte derfor et andet certifikat ind som tilfældigvis havde liggende (fra en anden testorganisation), og det kan være derfor at jeg får en fejl nu.

**Comment by martinwiboe** [comment]
Vil man ikke sende EntityId til STS, som indlejrer det i SAML-token, og som derefter udveksles til en Access Token? Således at dette token er knyttet til EntityId (eller måske blot indlejrer info om, hvilke privilegier, brugeren har).

Har du fulgt vejledningen i videoen (Episode 4) fra https://www.nemlog-in.dk/vejledningertiltestmiljo/ ?

Som jeg forstår det, skal certifikatet oprettes som et systemcertifikat via MitID Erhverv (og jeg har haft held med at få et sådant accepteret i NemLog-in administration ved oprettelse af IT-systembruger)

**Comment by erviem99** [comment]
Jeg har forsøgt at oprette en systemcertifikat sådan som der er beskrevet i videoen, men jeg får altid samme fejl tilbage i kaldet til STS'en: "The request was invalid or malformed".

**Comment by erviem99** [comment]
Digitaliseringsstyrelsen har lagt et nyt certifikat op her som virker: https://www.nemlog-in.dk/vejledningertiltestmiljo/forside/test.som.tjenesteudbyder.eller.broker/

De skriver 10. august, men det er altså først kommet på enten i går eller i dag.

Jeg kan desværre stadig ikke få det til at virke med vores egne systemcertifikater i testmiljøet :-(

**Comment by martinwiboe** [comment]
Jeg har ikke haft tid til at kigge mere på dette, men jeg har modtaget nedenstående fra NemLog-in support:

> Vi vil gerne informere dig om, at den aktuelle officielle sample, som du bruger, ikke fungerer i øjeblikket, da den kræver opdatering af certifikater. Dette er i tråd med behovet for at opdatere certifikaterne for alle Tjenesteudbydere. Vi er skiftet til brug af OCES3-certifikater i vores integrationstest, da de gamle sample-certifikater, OCES2, ophører med at fungere sammen med nedlukningen af NemID. Vi beklager ulejligheden, da Digitaliseringsstyrelsen (Digst) burde have informeret om dette på forhånd.
>
> I forhold til en opdateret version af den officielle sample, er vi opmærksomme på, at Digst har bestilt en opdatering. Men desværre er leveringstidspunktet for denne opdatering endnu ikke kendt.
>
> Det ser ud til, at client-certifikatet er i orden. Det er vigtigt at huske, at der er tre certifikater involveret i en OIO IDWS-integration, og det er muligt, at WSC-klienten skal opdatere et eller begge af de to andre certifikater, som også er blevet opdateret til OCES3.
>
> De tre nødvendige certifikater er:  
> 1. WSC client-certifikat: Dette er det certifikat, man logger ind med, og som er angivet i metadata. 
> 2. STS/IDP-certifikat: Dette certifikat bruges til at verificere signeringen fra STS, når WSC modtager sin assertion fra STS. (Jeg har > vedhæftet dette her: NemLog-in IdP - Test.cer)
> 3. WSP-certifikat: Dette certifikat bruges til at verificere signeringen fra LookupService, når man bruger REST-integration med AS-serveren. (Jeg har vedhæftet dette her: NemLog-in EIA.Frontend - Test.cer)
>
> Det er også vigtigt at installere trust til CA'erne.

**Comment by mikkeljohnsen** [comment]
@martinwiboe @erviem99 Fandt i ud af hvordan man kan bruge dette på .NET6 (eller nyere) til at tilgå CPR match opslagstjenesten ?

Jeg har igennen mange år prøvet at porte det til .NET6 ved at bruge System.ServiceModel.* NuGet pakkerne. Men jeg er ikke stærk nok i WS-Trust (siden det har været forældet i mange år).

**Comment by mikkeljohnsen** [comment]
Jeg ser nu selv på denne: https://services.test-devtest4-nemlog-in.dk/swagger/ui/index

Måske man ikke behøver SOAP

**Comment by martinwiboe** [comment]
@mikkeljohnsen jeg har ikke set mere på dette (andre opgaver kom i forkøbet) men jeg tror desværre at den API, du linker til, kræver et token som udstedes via SOAP.

**Comment by martinwiboe** [comment]
@mikkeljohnsen måske er denne tråd død, men jeg endte med at benytte ITfoxtec til at logge ind med MitID hos NemLog-in broker, og derefter SubjectMatchesCpr tjenesten, som fungerer via REST med et klientcertifikat. Der er et fint eksempel på https://tu.nemlog-in.dk/oprettelse-og-administration-af-tjenester/log.in/dokumentation-og-guides/

**Comment by mikkeljohnsen** [comment]
@martinwiboe Ja, jeg har også prøvet at bruge dette eksempel til at teste med. Men kan simpelthen ikke finde ud af hvor man skal oprette certifakater devtest4, test-devtest4 eller pp.mitid.dk.

Så mit næste forsøg er bare at teste i produktion, der er kun et sted til det.

Tak for info.

**Comment by martinwiboe** [comment]
Jeg fik det til at fungere ved at

- Oprette min tjeneste på https://administration.nemlog-in.dk (web login, med hak i kvalificeret signering)
- Oprette en testorganisation på i MitID simulatoren på https://testportal.test-devtest4-nemlog-in.dk/TU
- Udstede et systemcertifikat på https://erhvervsadministration.devtest4-nemlog-in.dk/certificates
- Angive entity id og uploade public key fra mit certifikat
- Benytte certifikatet som klientcertifikat ved kald til matchservice (jf. deres eksempelkode)

**Comment by mikkeljohnsen** [comment]
@martinwiboe 

"Oprette min tjeneste på https://administration.nemlog-in.dk/ (web login, med hak i kvalificeret signering)"

Er det "Log-in-tjeneste (privat) - OIOSAML 3" + "kvalificeret signering" ?

**Comment by martinwiboe** [comment]
Ja, lige præcis


------ Original Message ------
From "Mikkel Kruse Johnsen" ***@***.***>
To "digst/OIOIDWS.Net" ***@***.***>
Cc "Martin Wiboe" ***@***.***>; "Mention" 
***@***.***>
Date 30/01/2024 13.38.58
Subject Re: [digst/OIOIDWS.Net] Access denied in call (Issue #41)

**Comment by mikkeljohnsen** [comment]
@martinwiboe Beklager jeg lige spørger så meget, men har faktisk brugt meget tid på dette, uden det virkede.

Er det et "IT-System" du har tilføjet under "It-systemudbyder" ?

Eller er det bare en "Systembruger" som du har lavet ?

**Comment by erviem99** [comment]
Jeg endte også med REST løsningen da det var det eneste som fungerede. Den har vi også kørende i drift nu.

IT System er navnet på det IT system som brugeren er logget ind via.

Det personlige ID på brugeren i SAML tokenet er således bundet op mod IT Systemet og du kan kun validere CPR-nummeret mod det rigtige IT System.

**Comment by martinwiboe** [comment]
@mikkeljohnsen Ja, under IT-systemudbyder har jeg tilføjet systemet ("Log-in-tjeneste (privat) - OIOSAML 3" + "kvalificeret signering"). Jeg får adgang ved at logge ind med MitID Erhverv som min virksomhed.

Du skal så tilføje en teknisk administrator, som kan uploade certifikatet mv. Det sker via email, og ender med også at være bundet op på mit private MitID.
